presto-hive/src/main/java/com/facebook/presto/hive/s3/HiveS3Module.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.s3;
+
+import com.facebook.presto.hive.HiveClientConfig;
+import com.facebook.presto.hive.S3ConfigurationUpdater;
+import com.google.inject.Binder;
+import com.google.inject.Scopes;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
+import org.apache.hadoop.conf.Configuration;
+
+import static io.airlift.configuration.ConfigBinder.configBinder;
+import static java.util.Objects.requireNonNull;
+import static org.weakref.jmx.ObjectNames.generatedNameOf;
+import static org.weakref.jmx.guice.ExportBinder.newExporter;
+
+public class HiveS3Module
+        extends AbstractConfigurationAwareModule
+{
+    private static final String EMR_FS_CLASS_NAME = "com.amazon.ws.emr.hadoop.fs.EmrFileSystem";
+
+    private final String connectorId;
+
+    public HiveS3Module(String connectorId)
+    {
+        this.connectorId = requireNonNull(connectorId, "connectorId is null");
+    }
+
+    @Override
+    protected void setup(Binder binder)
+    {
+        if (buildConfigObject(HiveClientConfig.class).isUseEmrFileSystem()) {
+            binder.bind(S3ConfigurationUpdater.class).toInstance(HiveS3Module::setupEmrFileSystem);
+        }
+        else {
+            binder.bind(S3ConfigurationUpdater.class).to(PrestoS3ConfigurationUpdater.class).in(Scopes.SINGLETON);
+            configBinder(binder).bindConfig(HiveS3Config.class);
+
+            binder.bind(PrestoS3FileSystemStats.class).toInstance(PrestoS3FileSystem.getFileSystemStats());
+            newExporter(binder).export(PrestoS3FileSystemStats.class).as(generatedNameOf(PrestoS3FileSystem.class, connectorId));
+        }
+    }
+
+    private static void setupEmrFileSystem(Configuration config)
```

Comment:
Why is this done here for EmrFS, but in PrestoS3ConfigurationUpdater for PrestoS3FileSystem?

Reply:
This method is the implementation of the `S3ConfigurationUpdater` interface for EMRFS (it becomes an implementation when used as a lambda in the Guice binding call). But based on this conversation, we should pull it out into a real class to make it easier to read / more consistent.

Reply:
Actually, though, I'm wondering if we should split this out and verify that the class exists at config binding time, rather than waiting for someone to use it before failing. Thoughts?

Reply:
I think we could have it fail at config time as long as the exception messaging was sufficient.